silq/pulses/pulse_sequences.py

```diff
@@ -31,11 +37,626 @@ def __setattr__(self, key, value):
+            * ``ESR_pulses`` (List[Union[str, Pulse]]): List of ESR pulses to
+              use. Can be strings, in which case the string should be an item in
+              ``ESR`` whose value is a `Pulse`.
+            * ``pulse_delay`` (float): ESR pulse delay after beginning of stage
+              pulse. Default is 5 ms.
+            * ``read_pulse`` (Pulse): Pulse after stage pulse for readout and
+              initialization of electron. Default is 'read_initialize`
+              `DCPulse`.
+
+        EPR (dict): Pulse settings for the empty-plunge-read (EPR) part of the
+            pulse sequence. This part is optional, and is used for non-ESR
+            contast, and to measure dark counts and hence ESR contrast.
+            Contains the following items:
+
+            * ``enabled`` (bool): Enable EPR sequence.
+            * ``pulses`` (List[Pulse]): List of pulses for EPR sequence.
+              Default is ``empty``, ``plunge``, ``read_long`` `DCPulse`.
+
+        pre_pulses (List[Pulse]): Pulses before main pulse sequence.
+            Empty by default.
+        pre_ESR_pulses (List[Pulse]): Pulses before ESR readout pulse sequence.
+            Empty by default.
+        post_pulses (List[Pulse]): Pulses after main pulse sequence.
+            Empty by default.
+        pulse_settings (dict): Dict containing all pulse settings.
+        **kwargs: Additional kwargs to `PulseSequence`.
+
+    See Also:
+        NMRParameter
+
+    Notes:
+        For given pulse settings, `NMRPulseSequence.generate` will recreate the
+        pulse sequence from settings.
+    """
+    def __init__(self, pulse_sequences=None, **kwargs):
+        if pulse_sequences is None:
+            pulse_sequences = [
+                ElectronReadoutPulseSequence(name='NMR'),
+                ElectronReadoutPulseSequence(name='ESR')
+            ]
+
+        super().__init__(pulse_sequences=pulse_sequences, **kwargs)
+
+        self.ESR = next(pseq for pseq in self.pulse_sequences if pseq.name == 'ESR')
+        self.NMR = next(pseq for pseq in self.pulse_sequences if pseq.name == 'NMR')
+        try:
+            self.initialization = next(
+                pseq for pseq in self.pulse_sequences if pseq.name == 'initialization'
+            )
+        except:
+            self.initialization = None
+
+        # Disable read pulse acquire by default
+        self.NMR.settings['read_pulse'].acquire = False
+
+
+class T2PulseSequence(ElectronReadoutPulseSequence):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        self.pulse_settings.update({
+            'RF_initial_pulse': 0,
+            'RF_refocusing_pulse': 0,
+            'RF_final_pulse': 0,
+            'RF_inter_pulse': None,
+            'final_phase': 0,
+            'artificial_frequency': 0,
+            'num_refocusing': 0
+        })
+        self.tau: float = Parameter('tau', unit='s', initial_value=1e-3, parent=False)
+
```

Comment:
why did you create a separate T2 sequence? should the general ElectronReadoutSequence cover it?

Reply:
The T2PulseSequence is a subclass of ElectronReadoutPulseSequence. While in principle all its things can be done by the ElectronReadoutPulseSequence (it is a subclass after all), it has often-used T2 features grouped into settings, an example being `artificial_frequency`. This makes it easier to run a T2 pulse sequence

Reply:
Else you could also ask: why create the `ElectronReadoutSequence`, since the regular `PulseSequence` should cover it

Reply:
Understand, ok